### PR TITLE
Fix hasfiller value for BFiniteType created for const literals

### LINKED ISSUE #21251: Fix compile time fill member check suggestions for #20749
**Description:**
$title.

### PR DESCRIPTION
**Pull request by KRVPerera**
## Purpose
BFiniteType were compared using `==` operator in BSameTypeVisitor. This fails to identify that two BFiniteTypes belong to the same type (eg: `int`). Due to this reason sample code fails to compile. Failing condition is that a union must have a single basic type for all of its members to have a filler value.

Fixes #21251

## Approach
BFiniteTypes value space is flattened recursively and passed to BSameTypeVisitor.

## Samples
```Ballerina
const FOO = 0;
type Bar FOO|1;
public function main() {
    Bar[2] x = [];
}
```
List expression now supports this syntax..

https://github.com/ballerina-platform/ballerina-lang/issues/22403

```Ballerina
const FOO_ZERO = 0;
type FOO_FOUR_THREE  4 | 3;
type BarMultiple FOO_ZERO | 1 | FOO_FOUR_THREE;
function createMultipleConstLiteralAutoFilledSealedArray() {
    BarMultiple a = 1;
    BarMultiple[5] sealedArray = [a, a];
    sealedArray[3] = a;
    //assertEqualPanic(0, sealedArray[2]); 
}
```
The compiler still fails to fill arrays with valid values for complex unions. In above code array index `2` is not filled with any value.

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java [hunk 1]
```diff
@@ -2893,14 +2843,27 @@ private boolean checkFillerValue(BUnionType type) {
         if (type.isNullable()) {
             return true;
         }
-        Iterator<BType> iterator = type.getMemberTypes().iterator();
+        Set<BType> typeTags = new HashSet<>();
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java]: Type tags mean something else for the compiler right? Shall we use a different name?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/model/types/BFiniteType.java [hunk 2]
```diff
@@ -115,4 +116,16 @@ public void addValue(BLangExpression value) {
             nullable = true;
         }
     }
+
+    public Set<BType> getUniqueValues() {
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/model/types/BFiniteType.java]: The name of the function is not accurate right? 

Also, this is like a specific scenario so I think this shouldn't be a method in BFiniteType.java. We can maybe introduce a method in the Types.java class itself instead.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/model/types/BFiniteType.java [hunk 3]
```diff
@@ -115,4 +116,16 @@ public void addValue(BLangExpression value) {
...
+        Set<BType> uniqueType = new HashSet<>();
+        for (BLangExpression expression : this.valueSpace) {
+            if (expression.type.tag != TypeTags.FINITE) {
+                uniqueType.add(expression.type);
+            } else {
+                uniqueType.addAll(((BFiniteType) expression.type).getUniqueValues());
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/model/types/BFiniteType.java]: Can you please share an example where the member in the value space of a finite type is another finite type?

**Comment by MaryamZi** [comment]
Aren't we adding this to the 1.2.x branch btw? Couldn't find a PR to the ballerina-1.2.x branch. 

This doesn't break existing code right?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/model/types/BFiniteType.java (hunk 3 above)
Comment (KRVPerera) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/model/types/BFiniteType.java]: Think I have confused this with the union type

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java [hunk 4]
```diff
@@ -1887,18 +1843,15 @@ public Boolean visit(BServiceType t, BType s) {
...
                 return false;
             }
             BTypedescType sType = ((BTypedescType) s);
             return isSameType(sType.constraint, t.constraint, this.unresolvedTypes);
         }
 
-
         @Override
         public Boolean visit(BFiniteType t, BType s) {
-
```
[the remark this answers is not in the document]
Comment (KRVPerera) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java]: - Leave your code better than you found it. Boy Scout Rule.
- reverted.  I will check the guidelines.

**Comment by KRVPerera** [comment]
> Aren't we adding this to the 1.2.x branch btw? Couldn't find a PR to the ballerina-1.2.x branch.
> 
> This doesn't break existing code right?

I am waiting for the code review to finish.

**Pull request merged**